README.md

```diff
@@ -998,15 +998,50 @@ should also work if `oomph-lib` has already been configured, using a build direc
+***__Puneet__: Please check: No need to specify the flag during the build and install stages, right? ***
+```bash
+# Configure with range checking 
+cmake -G Ninja -B build -DOOMPH_ENABLE_RANGE_CHECKING=ON
+
+# Build the oomph-lib libraries 
+cmake --build build
+
+# Install
+cmake --install build
+``` 
+The table below provides a complete list of all variables and their default setting, typically chosen such that the fastest possible version of the library is built.
+
+The variables fall into several categories. There are those that get translated into C++ macros, such as 
+* `OOMPH_ENABLE_MPI` which enables `oomph-lib`'s MPI-based parallel capabilities via the C++ macro `OOMPH_HAS_MPI`. It also ensures that the sources code is compiled with a suitable MPI compiler *** __Puneet:__ I think you claimed that his was done automatically. If so, how? If not, does this require the simultaneous specification of mpi compilers for all three languages? **
+* `OOMPH_ENABLE_PARANOID` which enables the compilation of built-in self-tests and diagnostics via the C++ macro `PARANOID`. These tests are very useful during code development but add to the run-time
+* `OOMPH_ENABLE_RANGE_CHECKING`, already discussed above
 
 
+Other variables request that certain third-party libaries are built at the same time as `oomph-lib`. E.g. 
+* `OOMPH_WANT_MUMPS` requests that the parallel frontal solver MUMPS is built
+* `OOMPH_WANT_HYPRE` requests that the Hypre library is built
+* `OOMPH_WANT_TRILINOS` requests that the Trilinos library is built
+* *** __Puneet:__ what else? ***
 
+If any of these are set to `ON`, `CMake` will download the relevant sources and build the libraries as part of the `oomph-lib` installation. It will also (via associated C++ macros, `OOMPH_HAS_MUMPS`, `OOMPH_HAS_HYPRE`, `OOMPH_HAS_TRILINOS`) enable the compilation of certain `oomph-lib` wrappers to these libraries. For instance, if `OOMPH_WANT_MUMPS` is set to `ON` during the project configuration, `oomph-lib`'s C++ sources will be compiled with the C++ macro `OOMPH_HAS_MUMPS` which then enables the creation of a `LinearSolver` object that uses `MUMPS` as the underlying linear solver.
 
+Since third party libraries are not modified/maintained by us, it is, of course, not necessary to rebuild them over and over again. They only have to be installed once in a permanent location, or they may already be installed somewhere on your computer. In this case, their location can be specified via variables such as 
+* `OOMPH_USE_BLAS_FROM` which specifies the absolute path to the directory in which the BLAS library has been installed; this directory should contain an `include` and `lib` subdirectory *** __Puneet:__ Making this up as I go along; we ought to specify what exactly the path is to ***.
+* ** __Puneet:__ I assume there are others; list/describe here. **
```

Comment:
As before, the folders should contain an `include/` and `lib/` folder containing the relevant files for each library:

`OOMPH_USE_LAPACK_FROM`: same as `OOMPH_USE_BLAS_FROM`; doesn't need to specified if BLAS and LAPACK are the same library, e.g. for OpenBLAS
`OOMPH_USE_CGAL_FROM`: should NOT be used until CGAL patch their library and we can use patch release v5.5.2, but you can still specify your own Boost, GMP, MPFR installation which should decrease `oomph-lib` configure times
`OOMPH_USE_BOOST_FROM`: 1.66.0 or greater (for CGAL)
`OOMPH_USE_GMP_FROM`: v4.2 or greater (for CGAL)
`OOMPH_USE_MPRF_FROM`: v2.2.1 or greater (for CGAL)
`OOMPH_USE_MUMPS_FROM`: this relies on a CMake-based build of MUMPS; it won't work if you build from sources (yet!) as they use a Makefile-based build system which requires extra care
`OOMPH_USE_HYPRE_FROM`: should be v2.27.0 or greater
`OOMPH_USE_TRILINOS_FROM`: should be commit `0270d8bd4cd3c62fb37a30535f78ee16b9c77574` or later from the GitHub repository; they're moving the modern CMake practices and older versions make it very (very) difficult to integrate with the library as is. When they release v14.0 we can stick to specific versions
`OOMPH_USE_MPI_FROM`: for your MPI compilers/execs; haven't battle-tested this yet as CMake is able to find the files in the system paths for me **[NEED TO TEST]**

Reply:
Am keeping this open for now. Need to discuss in more detail. 

Also, why don't these not get listed when I do cmake -LH ?

Reply:
These aren't stored as cache variables; variables can be added to the cache through at least three ways:
1. specifying them from the commandline with `-D<var>=<value>`
2. by calling `set(<var> <value> CACHE <type> "<docstring"> [FORCE])`, and
3. by defining an `option(<var> <docstring> <value>)`
Currently, we just allow method (1). If these variables aren't specified, they won't be set anywhere so they won't appear in the CMake cache file (so they won't get listed by `cmake -LH`). When I add the `oomph_option(...)` and `oomph_set(...)` this won't be a problem anymore